Add implementation of MockGattServer::addService

pan- opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

It is not easy to write expectations for this function:
- The data in input is a complex tree that is not comparable.
- The data in input is modified: The attributes of each element of
the tree is added by the server.

To ease test, a default implementation has been added but it is
still possible to override it or set expectations.
That default implementation set unique attribute handles for
each element of the service and _copy_ it in the GattServerMock.

The copy does *not* follow standard BLE API, it is a _POD_ where
each element of the object is accessible as a field. Childs are
stored in vector to ease iteration and memory management.

The representation of the services registered into the mock are
accessible in GattServerMock::services.


### Documentation <!-- Required -->

Feature branch 

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [x] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

pan- committed:
Fix mbed_retarget.h stub on Mac OS.

Some parts of the stat structure are already defined by the system as macros.
The solution is to disable them. If these members are necessary for unit test,
a deeper integration is required, the best being to use the native structure.

pan- committed:
Add fake implementation for GattServerMock::addService.

It is not easy to write expectations for this function:
- The data in input is a complex tree that is not comparable.
- The data in input is modified: The attributes of each element of
the tree is added by the server.

To ease test, a default implementation has been added but it is
still possible to override it or set expectations.
That default implementation set unique attribute handles for
each element of the service and _copy_ it in the GattServerMock.

The copy does *not* follow standard BLE API, it is a _POD_ where
each element of the object is accessible as a field. Childs are
stored in vector to ease iteration and memory management.

The representation of the services registered into the mock are
accessible in GattServerMock::services.

paul-szczepanek-arm commented on UNITTESTS/target_h/platform/mbed_retarget.h:
```diff
@@ -374,9 +374,15 @@ struct stat {
 
     off_t     st_size;    ///< Size of file in bytes
 
+#ifndef st_atime
```
what is this? is this some macro hackery?

pan- replied:
Yes a macro `st_atime` is defined in MacOS. I can add a comment to make it more explicit.

paul-szczepanek-arm commented on UNITTESTS/fakes/ble/GattServerImpl_mock.h:
```diff
@@ -57,6 +57,47 @@ class GattServerMock : public ble::impl::GattServer {
+    };
+
+    // Characteristic representation of a characteristic registered with ble::test::register_services
+    struct characteristic_t {
```
why didn't you use the normal one?

pan- replied:
The main driver was getting access to the callbacks. but as you pointed out we can add a function for it. 

I started with using normal characteristics as a first prototype but managing the memory was painful because nothing is really owned by characteristics so you need a vector for the characteristics, another for their array, another for their value and it is the same thing for descriptors. It just makes the fake code much harder to read.

paul-szczepanek-arm reviewed (commented):
I'll wait until I see this in action in the unit test, for now I don't see the point yet

paul-szczepanek-arm commented on UNITTESTS/fakes/ble/source/GattServerImpl_mock.cpp:
```diff
@@ -0,0 +1,101 @@
+
+// Accessors to private callbacks of GattCharacteristic
+HIJACK_MEMBER(read_cb_access, FunctionPointerWithContext<GattReadAuthCallbackParams *> GattCharacteristic::*, &GattCharacteristic::readAuthorizationCallback);
+HIJACK_MEMBER(write_cb_access, FunctionPointerWithContext<GattWriteAuthCallbackParams *> GattCharacteristic::*, &GattCharacteristic::writeAuthorizationCallback);
```
I think a nicer solution would be to make public getters for these, there's not risk in a getter

paul-szczepanek-arm reviewed (dismissed):
I saw how it's used, looks good, maybe we can avoid adding hijack member and instead just have a public getter

pan- committed:
Fix fake event queue process event.

The result of remove_if was fed into erase without checking something
was actually removed.

pan- committed:
BLE: Add API to test the presence of an event handler in a chain.

pan- committed:
BLE: Add APIs to get characteristic authorization callbacks.
